### PR TITLE
Update start_index and end_index for adaptive pooling

### PR DESCRIPTION
**Pull request by Aidyn-A**
### Description
The PR fixes the issue #81409. To fix the issue the procedure of determining start and end indices for adaptive max pooling and average pooling is modified towards integer-only arithmetic. 

### Testing
The testing of the new functions is straightforward:

```
#include <iostream>
#include <cassert>
#include <cmath>

int64_t start_index(int64_t a, int64_t b, int64_t c) {
  return (int64_t)std::floor((float)(a * c) / b);
}

int64_t end_index(int64_t a, int64_t b, int64_t c) {
  return (int64_t)std::ceil((float)((a + 1) * c) / b);
}


int64_t start_index_new(int64_t a, int64_t b, int64_t c) {
  return (a / b) * c + ((a % b) * c) / b;
}

int64_t end_index_new(int64_t a, int64_t b, int64_t c) {
  return 1 + ((a + 1) * c - 1) / b;
}


int main() {
    size_t N = 2<<24;
    std::cout<<N<<'\n';
    int64_t c = 1;

    for(int64_t i=1; i<N; i++) {
        for(int64_t j=1; j<N; j++) {
            int64_t s_id0 = start_index(i, j, c);
            int64_t s_id1 = start_index_new(i, j, c);
            assert(s_id0 == s_id1);
        }
    }

    for(int64_t i=1; i<N; i++) {
        for(int64_t j=1; j<N; j++) {
            int64_t e_id0 = end_index(i, j, c);
            int64_t e_id1 = end_index_new(i, j, c);
            assert(e_id0 == e_id1);
        }
    }
}
```

### CODE REVIEW

File: test/nn/test_pooling.py [hunk 1]
```diff
@@ -1421,6 +1421,18 @@ def test_adaptive_pool_invalid(self, device):
                 t, output_size = inp
                 m(output_size)(t)
 
+    @slowTest
+    def test_adaptive_pool_odd_size(self, device):
+        # See https://github.com/pytorch/pytorch/issues/81409
+        Ih, Iw, Oh, Ow = 5873, 3693, 3527, 2219
```
Comment (Aidyn-A) [inline test/nn/test_pooling.py:1427]: These are specifically chosen numbers that are causing an old version to segfault in #81409

File: aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu [hunk 2]
```diff
@@ -28,12 +28,12 @@ namespace native {
 
 namespace {
 
-__device__ inline int start_index(int a, int b, int c) {
-  return (int)std::floor((float)(a * c) / b);
+__device__ inline int64_t start_index(int64_t a, int64_t b, int64_t c) {
+  return (a * c) / b;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu:32]: If I understand correctly, you could still overflow the a * c multiplication here, right? I wonder if we should do something like in https://stackoverflow.com/questions/5542037/avoiding-overflow-in-integer-multiplication-followed-by-division

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu (hunk 2 above)
Comment (Aidyn-A) [reply aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu:32]: That is a good suggestion, I will implement it. Thanks!

**Pull request closed without merging**